### PR TITLE
Added Parallelism configuration option to lakectl

### LINKED ISSUE #8159: Allow for the lakectl local --parallelism flag to have persistent default values through env vars/config
This ticket requests that a mechanism is provided to set a default value of the `--parallelism` flag for the `lakectl local` commands. Currently, it defaults to 25 and must be explicitly specified on the command line in order to change the default.

It would be fantastic to allow an environment variable and/or config file to set the default value here, similarly to how some of the configs here: https://docs.lakefs.io/reference/configuration.html influence the local CLI, and allow for [use of env vars](https://docs.lakefs.io/reference/configuration.html#using-environment-variables) to set them.

Motivation: I have seen a big performance improvement in `lakectl local` when working with lots of smaller files with an increased `--parallelism` flag. Rather than needing to remember to set this flag each time, wrap the tool in helper scripts, etc. it would be great if there was a way to provide a higher custom default directly to the tool.

First asked about here: https://lakefs.slack.com/archives/C016726JLJW/p1725922712803829

### PR DESCRIPTION
**Pull request by ItamarYuran**
Closes #8159

## Change Description
Added the option to configure the amount of parallelism for lakectl.

See [feature request](https://github.com/treeverse/lakeFS/issues/8159)  for more backround.

Currently, the configuration nests in a new section "Options"

### CODE REVIEW

File: cmd/lakectl/cmd/root.go [hunk 1]
```diff
@@ -153,9 +156,10 @@ const (
 	parallelismFlagName   = "parallelism"
 	noProgressBarFlagName = "no-progress"
 
-	defaultSyncParallelism = 25
-	defaultSyncPresign     = true
-	defaultNoProgress      = false
+	defaultParallelismConfig = -1
```
Comment (arielshaqed) [inline cmd/lakectl/cmd/root.go:159]: Can we use 0?  If not, we need to document what "-1" means.

File: pkg/local/sync.go [hunk 2]
```diff
@@ -74,7 +74,11 @@ func (s *SyncManager) Sync(rootPath string, remote *uri.URI, changeSet <-chan *C
 	defer s.progressBar.Stop()
 
 	wg, ctx := errgroup.WithContext(s.ctx)
-	for i := 0; i < s.cfg.SyncFlags.Parallelism; i++ {
+	parallelismToUse := s.cfg.Parallelism
+	if parallelismToUse <= 0 {
+		parallelismToUse = s.cfg.SyncFlags.Parallelism
+	}
+	for i := 0; i < parallelismToUse; i++ {
```
Comment (arielshaqed) [inline pkg/local/sync.go:81]: 2 changes, the first is more important:

- AFAIU SyncFlags.Parallelism is more specific than the new setting, so if both are specified we should use it and not the new more global setting.
- It is enough to call the var "parallelism", it will obviously be used.
```suggestion
	parallelism := s.cfg.SyncFlags.Parallelism
	if parallelism <= 0 {
		parallelism = s.cfg.Parallelism
	}
	for i := 0; i < parallelism; i++ {
```

**Review by arielshaqed — CHANGES_REQUESTED** [review]
Thanks!  Concise fix :-)

Confused about the precedence of the 2 settings for parallelism, so asking for a change in case I am right about the required precedence.

File: pkg/local/sync.go (hunk 2 above)
Comment (ItamarYuran) [reply pkg/local/sync.go:81]: Thank you!
I understand, problem is,  the default for `s.cfg.SyncFlags.Parallelism` is 25, so unless the user flagged with a 0 or less, the used value will be the default flag and not the configuration.

File: cmd/lakectl/cmd/root.go [hunk 3]
```diff
@@ -153,9 +156,10 @@ const (
 	parallelismFlagName   = "parallelism"
 	noProgressBarFlagName = "no-progress"
 
-	defaultSyncParallelism = 25
-	defaultSyncPresign     = true
-	defaultNoProgress      = false
+	defaultParallelismConfig = 25
+	defaultSyncParallelism   = 0
```
Comment (N-o-Z) [inline cmd/lakectl/cmd/root.go:160]: I'm not sure why both are needed.
We should:
1. Use configuration param if provided
2. Use modified param value if passed as part of command
3. Fallback to default param value

**Review by N-o-Z — CHANGES_REQUESTED** [review]
I'm not sure I understand the flag redundancy here.
You can have both a default param value and override it with a config value if exists it's not mutually exclusive.

File: cmd/lakectl/cmd/root.go (hunk 3 above)
Comment (ItamarYuran) [reply cmd/lakectl/cmd/root.go:160]: My thought was, that positive default value should come from one source only. In case both have positive default values how will we determine which one to use? I chose the config val to be positive by default, following Ariel's [change request](https://github.com/treeverse/lakeFS/pull/8283#discussion_r1798350158) to go from the more general case to specific case.
In any case, this code run according to the priorities:
- no flag no config val - uses 25 as config default 
- no flag with config - uses config
- flag with no config - uses the flag, 
-  flag with config - uses flag

Comment (N-o-Z) [reply cmd/lakectl/cmd/root.go:160]: Your priorities seem correct. We have a way to determine whether a flag in Cobra was explicitly provided by the user or not (there's at least one place in the code where we use it AFAIR).
You can use the same method to do that and determine whether we need to use the config value or the param value

File: pkg/local/config.go [hunk 4]
```diff
@@ -17,6 +18,7 @@ type SyncFlags struct {
 
 type Config struct {
 	SyncFlags
+	Parallelism int
```
Comment (N-o-Z) [inline pkg/local/config.go:21]: That's not needed anymore, right?

File: pkg/local/sync.go [hunk 5]
```diff
@@ -74,7 +74,11 @@ func (s *SyncManager) Sync(rootPath string, remote *uri.URI, changeSet <-chan *C
 	defer s.progressBar.Stop()
 
 	wg, ctx := errgroup.WithContext(s.ctx)
-	for i := 0; i < s.cfg.SyncFlags.Parallelism; i++ {
+	parallelism := s.cfg.Parallelism
```
Comment (N-o-Z) [inline pkg/local/sync.go:77]: This change is not needed anymore?

File: cmd/lakectl/cmd/root.go [hunk 6]
```diff
@@ -245,9 +248,11 @@ func getSyncFlags(cmd *cobra.Command, client *apigen.ClientWithResponses) local.
 	if parallelism < 1 {
 		DieFmt("Invalid value for parallelism (%d), minimum is 1.\n", parallelism)
 	}
+	setParallelism := cmd.Flags().Changed(parallelismFlagName)
 
 	presignMode := getPresignMode(cmd, client)
 	return local.SyncFlags{
+		SetParallelism:   setParallelism,
```
Comment (N-o-Z) [inline cmd/lakectl/cmd/root.go:255]: You don't really need SetParallelism anymore.
Note that you have cfg in your context here, so basically you can make the decision (what value to use for the parallelism) here.

**Review by N-o-Z — CHANGES_REQUESTED** [review]
@ItamarYuran thanks! I think we're in the right direction!
Added some more comments, if anything is not clear - let's sync to reduce the review cycle.

File: pkg/local/config.go [hunk 7]
```diff
@@ -9,6 +9,7 @@ const (
 )
 
 type SyncFlags struct {
+	SetParallelism   bool
```
Comment (N-o-Z) [inline pkg/local/config.go:12]: That's not needed anymore, right?

File: cmd/lakectl/cmd/root.go [hunk 8]
```diff
@@ -574,6 +579,7 @@ func initConfig() {
 	viper.SetDefault("server.retries.min_wait_interval", defaultMinRetryInterval)
 	viper.SetDefault("experimental.local.posix_permissions.enabled", false)
 	viper.SetDefault("local.skip_non_regular_files", false)
+	viper.SetDefault("options.parallelism", defaultParallelism)
```
Comment (N-o-Z) [inline cmd/lakectl/cmd/root.go:582]: You don't need a default for this value since the default is handled in param

**Comment by ItamarYuran** [comment]
@N-o-Z Thank you for your review!
Got rid of all the redundancies. LMK what do you think.

File: pkg/local/sync.go [hunk 9]
```diff
@@ -72,9 +72,10 @@ func NewSyncManager(ctx context.Context, client *apigen.ClientWithResponses, htt
 func (s *SyncManager) Sync(rootPath string, remote *uri.URI, changeSet <-chan *Change) error {
 	s.progressBar.Start()
 	defer s.progressBar.Stop()
-
 	wg, ctx := errgroup.WithContext(s.ctx)
-	for i := 0; i < s.cfg.SyncFlags.Parallelism; i++ {
+	parallelism := s.cfg.SyncFlags.Parallelism
+
+	for i := 0; i < parallelism; i++ {
```
Comment (N-o-Z) [inline pkg/local/sync.go:78]: Can you please revert this change (redundant assignment)

**Review by N-o-Z — APPROVED** [review]
@ItamarYuran Perfect!
Please note the single comment remained

**Review by arielshaqed — APPROVED** [review]
Thanks for the changes, and thanks for continuing the review @N-o-Z in the best way!  @N-o-Z 's comment remains, of course.
